### PR TITLE
kmeans clustering

### PR DESCRIPTION
**Pull request by ashish-1600**
### Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### Description

1. Added a new operator for clustering. 
2. Added a new operator for finding a common topic across different queries.

### Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/uptrain-ai/uptrain/blob/main/CONTRIBUTING.md) document.
- [x] My code follows the code style (BLACK) of this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have updated the documentation accordingly.
- [ ] I have added an appropriate CHANGELOG entry.

### Author's Note

Please provide any additional information, questions, or concerns you have about this pull request.

### CODE REVIEW

File: uptrain/operators/language/llm.py
Comment (Dominastorm) [inline uptrain/operators/language/llm.py:1]: Hey Ashish, can you explain why you changed this function?

File: uptrain/operators/language/llm.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ async def async_process_payload(
         ):  # failed requests don't count towards rate limit
             try:
                 if payload.data["model"].startswith("gpt"):
-                    payload.response = await openai.ChatCompletion.acreate(
+                    payload.response = await openai.ChatCompletion.create(
```
Comment (Dominastorm) [inline uptrain/operators/language/llm.py:53]: Hey Ashish, can you explain why you changed this function?

Comment (ashish-1600) [reply uptrain/operators/language/llm.py:53]: Hey Dhruv, While doing development, was exploring some change here, missed to change it back.. Thanks for pointing it out.. Pushing another commit to revert this change.

Comment (Dominastorm) [reply uptrain/operators/language/llm.py]: No problem. Can you do the same for lines 126, 135 and 137 in uptrain/operators/language/generation.py?

Comment (ashish-1600) [reply uptrain/operators/language/llm.py]: Thanks for pointing it out. Pushed another commit for the same.

File: uptrain/operators/clustering.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
+"""
+This module provides the following Distribution and UMAP operators that operate on embeddings:
```
Comment (sourabhagr) [inline uptrain/operators/clustering.py:2]: Let's change the description @ashish-1600

File: uptrain/operators/clustering.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    def setup(self, settings: Settings):
+        if self.algorithm == "kmeans":
+            self.algorithm_obj = nltk.cluster.KMeansClusterer(self.n_clusters, distance=nltk.cluster.util.cosine_distance,avoid_empty_clusters=True)
+        else:
+            raise Exception(f"{self.algorithm} is not supported yet.")
+        return self
+
+    def run(self, data: pl.DataFrame) -> TYPE_TABLE_OUTPUT:      
+        embs_1 = np.asarray(data[self.col_in])
```
Comment (sourabhagr) [inline uptrain/operators/clustering.py:77]: better variable name than embs_1 @ashish-1600

File: uptrain/operators/embs.py [hunk 4]
```diff
@@ -176,52 +179,27 @@ class UMAP(TransformOp):
...
+                [0.072772, 0.073564, … -0.043947]
+        ]
         ```
 
     """
 
-    col_in_embs_1: str
-    col_in_embs_2: str
+    col_in_embs: str = 'embedding'
+    n_components: int
```
Comment (sourabhagr) [inline uptrain/operators/embs.py:193]: We can give a default value for n_components @ashish-1600

File: uptrain/operators/clustering.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        # Run the operator on the input data
+        input_data = pl.DataFrame(...)
+        output = op.run(input_data)
+
+        # Get the output DataFrame
+        assigned_clusters = output["output"]
+        ```
+    """
+
+    algorithm: str
```
Comment (sourabhagr) [inline uptrain/operators/clustering.py:64]: Default values for algorithm and num_cluster @ashish-1600

File: uptrain/operators/language/generation.py [hunk 6]
```diff
@@ -170,6 +171,104 @@ def run(self, data: pl.DataFrame) -> TYPE_TABLE_OUTPUT:
...
+        col_in_embs (str): The name of the column containing the embeddings.
+        col_in_text (str): The name of the column containing the text where the grouping needs to be performed.
+        col_out_text (str): The name of the column containing the topic for each entry.
+        temperature (float): Temperature for the LLM to generate responses.
+        model (str): which LLM to use for topic generator.
+
+    Returns:
+        TYPE_TABLE_OUTPUT: A dictionary containing the dataset with the output text.
+    """
+    col_in_cluster_index: str = 'cluster_index'
```
Comment (sourabhagr) [inline uptrain/operators/language/generation.py:190]: We should restructure it to assume clustering as an optional? Also top5 -> 5 should be a parameter @ashish-1600

File: uptrain/operators/language/generation.py [hunk 7]
```diff
@@ -170,6 +171,104 @@ def run(self, data: pl.DataFrame) -> TYPE_TABLE_OUTPUT:
...
+        outputs = [None] * len(questions)
+        indexes_cluster = []
+        
+        for index in range(n_clusters+1):
+            points = data[self.cluster_index]==index
+            indexes_cluster.append(list(np.where(points)[0]))
+            embeddings_cluster = embeddings[points]
+            questions_cluster = questions[points]
+            centroid = np.mean(embeddings_cluster, axis=0)
+            indexes_sorted = np.argsort(np.linalg.norm(embeddings_cluster - centroid, axis=1))
```
Comment (sourabhagr) [inline uptrain/operators/language/generation.py:228]: This might lead to duplicates. Should we do sampling? @ashish-1600

Comment (ashish-1600) [reply uptrain/operators/language/generation.py]: Sampling might give too much variability. Should we take the top ones which are unique? @sourabhagr

**Review by sourabhagr — APPROVED** [review]

File: uptrain/operators/language/generation.py (hunk 7 above)
Comment (sourabhagr) [reply uptrain/operators/language/generation.py]: I guess we can allow both. My thoughts:
1. Top K uniques -> We might end up with very similar-looking candidates and that might not reflect the true topic of the cluster
2. Sampling -> Agreed, can bring variability across runs. 

Will raise an issue for this.

**Pull request merged**